src/ast/passes/resource_analyser.cpp

```diff
@@ -359,6 +364,13 @@ void ResourceAnalyser::visit(Call &call)
     // and symbols resolved even when unavailable at resolution time
     resources_.probes_using_usym.insert(probe_);
   }
+
+  if (call.func == "print" || call.func == "printf" || call.func == "time" ||
+      call.func == "unwatch" || call.func == "join" || call.func == "exit" ||
+      call.func == "clear" || call.func == "zero" || call.func == "system" ||
+      call.func == "cat") {
```

Comment:
Ideally this is an attribute in the function registry that applies to builtins? It doesn't need to happen at this time, but it would be good to think about incrementally moving towards a strategy where data on the builtins isn't scattered and hard-coded ;)

Reply:
> moving towards a strategy where data on the builtins isn't scattered and hard-coded ;)

💯 - It pained me to write this if statement.

Reply:
This list seems risky and feels complicated. There's no real runtime cost to having an unused single entry 8 byte map, right? I doubt any non-trivial programs will be missing all of these, so we'll have added a bunch of conditionals and complexity for not much. So what is this commit optimizing for?

Reply:
I think this is a fair comment and TBH I was originally skeptical of submitting this PR because it didn't seem worth the tradeoff. I might still be in that camp but I also think that even though this conditional is ugly, it should be fairly safe as you get a nice verifier error or segfault if you try to reference this map in codegen and it doesn't exist so any runtime tests would catch this.

I'm fine to abandon and make a smaller PR just moving the recursion entry check to the resource analyser but I also think it's nice getting rid of unnecessary codegen (like unused maps).

Reply:
>  we should probably not land the first commit

Fine with me - this is not a hill I care to die on.

>  Recursion analysis is only a codegen concern, right? Would it make more sense to throw it into codegen_llvm.cpp? Or codegen_helper.cpp?

Let me try to find a good home for it. Maybe I'll wait until after @amscanne 's pass refactor lands.